File: src/main/java/me/pcasaes/hexoids/infrastructure/kafka/converter/EventDtoDeserializer.java
package me.pcasaes.hexoids.infrastructure.kafka.converter;

import com.google.protobuf.InvalidProtocolBufferException;
import org.apache.kafka.common.serialization.Deserializer;
import org.apache.kafka.common.serialization.StringDeserializer;
import pcasaes.hexoids.proto.Event;

import java.util.Map;
import java.util.logging.Level;
import java.util.logging.Logger;

import static me.pcasaes.hexoids.domain.model.DtoUtils.EVENT_THREAD_SAFE_BUILDER;

public class EventDtoDeserializer implements Deserializer<Event> {

    private static final Logger LOGGER = Logger.getLogger(EventDtoDeserializer.class.getName());
    private final StringDeserializer stringDeserializer = new StringDeserializer();


    @Override
    public void configure(Map<String, ?> configs, boolean isKey) {
        stringDeserializer.configure(configs, isKey);
    }

    @Override
    public Event deserialize(String topic, byte[] data) {
        if (data == null) {
            return null;
        }
        try {
            return EVENT_THREAD_SAFE_BUILDER
                    .get()
                    .clear()
                    .mergeFrom(data)
                    .build();
        } catch (InvalidProtocolBufferException ex) {
            LOGGER.log(Level.SEVERE, ex.getMessage(), ex);
        }
        return null;

    }

    @Override
    public void close() {
        stringDeserializer.close();
    }
}

File: src/main/kotlin/me/pcasaes/hexoids/infrastructure/kafka/converter/EventDtoDeserializer.kt
package me.pcasaes.hexoids.infrastructure.kafka.converter

import com.google.protobuf.InvalidProtocolBufferException
import org.apache.kafka.common.serialization.Deserializer
import org.apache.kafka.common.serialization.StringDeserializer
import pcasaes.hexoids.proto.Event
import java.util.logging.Level
import java.util.logging.Logger

class EventDtoDeserializer : Deserializer<Event?> {
    private val stringDeserializer = StringDeserializer()


    override fun configure(configs: MutableMap<String, *>, isKey: Boolean) {
        stringDeserializer.configure(configs, isKey)
    }

    override fun deserialize(topic: String, data: ByteArray?): Event? {
        if (data == null) {
            return null
        }
        try {
            return Event.newBuilder()
                .mergeFrom(data)
                .build()
        } catch (ex: InvalidProtocolBufferException) {
            LOGGER.log(Level.SEVERE, ex.message, ex)
        }
        return null
    }

    override fun close() {
        stringDeserializer.close()
        super.close()
    }

    companion object {
        private val LOGGER: Logger = Logger.getLogger(EventDtoDeserializer::class.java.getName())
    }
}
